File: Data Structure/2018102111_서보민_lab05/3번/QueType.cpp
#include "QueType.h"

QueType<int>::QueType(int max)
// Parameterized class constructor
// Post: maxQue, front, and rear have been initialized.
//       The array to hold the queue elements has been dynamically
//       allocated.
{
  maxQue = max + 1;
  front = maxQue - 1;
  rear = maxQue - 1;
  items = new ItemType[maxQue];
}
QueType<int>::QueType()          // Default class constructor
// Post: maxQue, front, and rear have been initialized.
//       The array to hold the queue elements has been dynamically
//       allocated.
{
  maxQue = 501;
  front = maxQue - 1;
  rear = maxQue - 1;
  items = new ItemType[maxQue];
}
QueType<int>::~QueType<int>()         // Class destructor
{
  delete [] items;
}

void QueType<int>::MakeEmpty()
// Post: front and rear have been reset to the empty state.
{
  front = maxQue - 1;
  rear = maxQue - 1;
}

bool QueType<int>::IsEmpty() const
// Returns true if the queue is empty; false otherwise.
{
  return (rear == front);
}

bool QueType<int>::IsFull() const
// Returns true if the queue is full; false otherwise.
{
  return ((rear + 1) % maxQue == front);
}

void QueType<int>::Enqueue(ItemType newItem)
// Post: If (queue is not full) newItem is at the rear of the queue;
//       otherwise a FullQueue exception is thrown.  
{
  if (IsFull())
    throw FullQueue();
  else
  {
    rear = (rear +1) % maxQue;
    items[rear] = newItem;
  }
}

void QueType<int>::Dequeue(ItemType& item)
// Post: If (queue is not empty) the front of the queue has been 
//       removed and a copy returned in item; 
//       othersiwe a EmptyQueue exception has been thrown.
{
  if (IsEmpty())
    throw EmptyQueue();
  else
  {
    front = (front + 1) % maxQue;
    item = items[front];
  }
}

bool QueType<int>::Identical(QueType<ItemType>& queue)
{
    ItemType item, item1;
    while (!this->IsEmpty() && !queue.IsEmpty()) {
        this->Dequeue(item);
        queue.Dequeue(item1);
        if (item != item1)
            return false;
    }
    return true;
}

File: Data Structure/2018102111_서보민_PP06/QueLL/QueType.py
class NodeType:
    """ Node Type """
    def __init__(self, item):
        self.info = item
        self.next = None

class QueType:
    def __init__(self):
        self.front = None
        self.rear = None

    def make_empty(self):
        while self.front != None:
            tempPtr = self.front
            self.front = self.front.next
            del tempPtr
        self.rear = None

    def is_full(self):
        try:
            location = NodeType("test")
            del location
            return False
        except:
            return True

    def is_empty(self):
        return self.front == None

    def enqueue(self, item):
        if self.is_full() == True:
            print("Full queue")
        else:
            node = NodeType(item)
            node.next = None
            if self.rear == None:
                self.front = node
            else:
                self.rear.next = node
            self.rear = node

    def dequeue(self):
        if self.is_empty() == True:
            print("Empty Queue")
        else:
            tempPtr = self.front
            item = self.front.info
            self.front = self.front.next
            if self.front == None:
                self.rear = None
            del tempPtr
            return item


        
